src/asl.ts

```diff
@@ -1516,6 +1535,72 @@ export class ASL {
           return this.eval(values.expr);
         }
         throw new SynthError(ErrorCodes.Unsupported_Use_of_Promises);
+      } else if (
+        isPropAccessExpr(expr.expr) &&
+        isIdentifier(expr.expr.expr) &&
+        expr.expr.expr.name === "JSON"
+      ) {
+        if (expr.expr.name === "stringify") {
+          const heap = this.newHeapVariable();
+
+          const objParamExpr = expr.args[0].expr;
+          if (!objParamExpr || isUndefinedLiteralExpr(objParamExpr)) {
+            // return an undefined variable
+            return {
+              jsonPath: heap,
+            };
+          }
+
+          return this.evalExpr(
+            objParamExpr,
+            (_, { normalizeOutputToJsonPath }) => {
+              const objectPath = normalizeOutputToJsonPath().jsonPath;
```

Comment:
So, if i understand correclty, `normalizeOutputToJsonPath` is a way for the evaluator to "pull forward" the value as a JSON path only when needed. What happens if it is never called? Does the value not get evaluated and stored on the heap?

Reply:
Its tied to the state returned, so it will be executed immediately before the operation (ex: stringify).

When we write the expression handlers, we ensure the heap value is only created as needed. In the cases with short circuiting, we do not hoist to the top of the expression, we localization the heap assignment. This is why the consuming expression determines how to use the output rather than the expression that generates the output. Sometimes the output is ignored (ExprStmt), sometimes its used as is (lambda input), sometimes it needs to be normalized (JSON.stringify).

```ts
JSON.stringify({ a: "a" });
```

result (as ts because I don't want to write out ASL)

```ts
const heap1 = { a: "a" } // an object literal cannot be an input to States.JsonToString, need a json path
const heap2 = { string: JSON.stringify(heap1) } // ugly, but we do this because step functions does not support intrinsic functions uniformly. (can't be in inputpath, result path, etc).
return heap2.string; // return a reference to the resulting string
```

```ts
x ? JSON.stringify({ a: "a" }) : JSON.stringify({ b: "b" });
```

result (as ts because I don't want to write out ASL)

```ts
let temp;
if(x) {
   const heap1 = { a: "a" } // an object literal cannot be an input to States.JsonToString, need a json path
   const heap2 = { string: JSON.stringify(heap1) } // ugly, but we do this because step functions does not support intrinsic functions uniformly. (can't be in inputpath, result path, etc).
   temp = heap2.string; // return a reference to the resulting string
} else {
   const heap3 = { b: "b" } // an object literal cannot be an input to States.JsonToString, need a json path
   const heap4 = { string: JSON.stringify(heap3) } // ugly, but we do this because step functions does not support intrinsic functions uniformly. (can't be in inputpath, result path, etc).
   temp = heap4.string; // return a reference to the resulting string
}
```